lib/src/utils/concurrent_hashmap.rs

```diff
@@ -0,0 +1,185 @@
+use std::collections::HashMap;
+use std::hash::Hash;
+use std::sync::{Arc, RwLock};
+
+pub struct ConcurrentHashMap<K, V> {
+    inner: Arc<RwLock<HashMap<K, V>>>,
+}
+
+#[allow(dead_code)]
+impl<K, V> ConcurrentHashMap<K, V>
+where
+    K: Eq + Hash + Clone,
+    V: Clone,
+{
+    pub fn new() -> Self {
+        Self {
+            inner: Arc::new(RwLock::new(HashMap::new())),
+        }
+    }
+
+    pub fn with_capacity(capacity: usize) -> Self {
+        Self {
+            inner: Arc::new(RwLock::new(HashMap::with_capacity(capacity))),
+        }
+    }
+
+    pub fn insert(&self, key: K, value: V) -> Option<V> {
+        let mut map = self.inner.write().unwrap();
+        map.insert(key, value)
+    }
+
+    pub fn get(&self, key: &K) -> Option<V> {
+        let map = self.inner.read().unwrap();
+        map.get(key).cloned()
+    }
```

Comment:
We can do the same Borrow shenanigans from the stdlib map to avoid having to create owned strings for lookups:

```suggestion
    pub fn get<Q>(&self, key: &Q) -> Option<V>
    where
        K: Borrow<Q>,
        Q: Hash + Eq + ?Sized,
    {
        let map = self.inner.read().unwrap();
        map.get(key).cloned()
    }
```

Same for remove and contains as well